crates/agglayer-certificate-orchestrator/src/lib.rs

```diff
@@ -151,11 +210,49 @@ where
     }
 }
 
-impl<C, A, I> Future for CertificateOrchestrator<C, A, I>
+impl<C, E, A, I, P> CertificateOrchestrator<C, E, A, I, P>
 where
-    I: Send + Unpin + 'static,
+    I: Send + Unpin + 'static + CertificateInput,
+    P: Send + Unpin + 'static,
+    A: Certifier<Input = I, Proof = P>,
+    E: EpochPacker<Item = P>,
+{
+    fn receive_proof(&mut self, certifier_outputs: CertifierOutput<P>) {
+        let network_id = certifier_outputs.network;
+        self.pending_proofs
+            .entry(network_id)
+            .or_default()
+            .push_back(certifier_outputs.proof);
+
+        self.global_state
+            .entry(network_id)
+            .and_modify(|s| *s = certifier_outputs.new_state);
+    }
+
+    fn receive_certificate(&mut self, certificate: I) -> Result<(), Error> {
+        let network: NetworkId = certificate.network_id();
+        let local_state = self.global_state.entry(network).or_default().clone();
+
+        self.received_certificates
+            .entry(network)
+            .or_default()
+            .push_back(certificate.clone());
+
+        let task = self.certifier_task_builder.clone();
+        self.certifier_tasks
+            .spawn(async move { task.certify(local_state, certificate)?.await });
```

Comment:
Minor: Could this be simplified a bit by removing the layer of async/await?

```suggestion
        self.certifier_tasks
            .spawn(task.certify(local_state, certificate)?);
```

Reply:
hm it seems not possible as is in the current state because of lifetime issues